java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcDateVectorAccessor.java

```diff
@@ -108,11 +104,36 @@ private void fillHolder() {
 
   @Override
   public Timestamp getTimestamp(Calendar calendar) {
-    Date date = getDate(calendar);
-    if (date == null) {
+    final LocalDateTime localDateTime = getLocalDateTime(calendar);
+    if (localDateTime == null) {
+      return null;
+    }
+
+    return Timestamp.valueOf(localDateTime);
+  }
+
+  private LocalDateTime getLocalDateTime(Calendar calendar) {
+    getter.get(getCurrentRow(), holder);
+    this.wasNull = holder.isSet == 0;
+    this.wasNullConsumer.setWasNull(this.wasNull);
+    if (this.wasNull) {
       return null;
     }
-    return new Timestamp(date.getTime());
+
+    final LocalDateTime localDateTime =
+        DateUtility.getLocalDateTimeFromEpochMilli(this.timeUnit.toMillis(holder.value));
+    final ZoneId defaultTimeZone = Calendar.getInstance().getTimeZone().toZoneId();
```

Comment:
I don't think we want to use the system locale? It should just be UTC (even that is strictly wrong; the date is unzoned by default and only zoned when a calendar comes into play)

Reply:
Or does JDBC expect that it will be in the system timezone?

Reply:
Hi, problem here is that `java.sql.Timestamp` and `java.sql.Date` only contains a long value with milliseconds. You have to tell java how to print this long, in what timezone.

So, the method has to follow the same approach done at `org.apache.arrow.driver.jdbc.accessor.impl.calendar.ArrowFlightJdbcTimeStampVectorAccessor#getLocalDateTime`

- first putting the received value (which is in UTC) at the expected time zone. The expected time zone is the default one  for the JVM or the time zone specified when calling to `java.sql.ResultSet#getDate(java.lang.String, java.util.Calendar)`. This is done on this line.
 `.atZone(sourceTimeZone)`
- then, the value is adjusted to the JVM timezone. 
 `.withZoneSameInstant(defaultTimeZone)` 

When this method is used  `java.sql.ResultSet#getDate(java.lang.String, java.util.Calendar)`, it is expected that the caller prints the date returned on the specified time zone. Something like:
```
final Date dateValue = rs.getDate("date_field", Calendar.getInstance(TimeZone.getTimeZone("Asia/Tokyo")));
final SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
sdf.setTimeZone(TimeZone.getTimeZone("Asia/Tokyo"));
System.out.println("Received date: " +sdf.format(dateValue));
```

Reply:
Sorry, I found a little time to look into this finally.

I still don't see why the system timezone has to come into play.

Arrow date is with reference to the UNIX epoch, i.e. UTC. Java Date is with reference to GMT. As the Java docs themselves state, they are technically different but practically the same.

(1) If there is no calendar given, I would expect there to be no datetime arithmetic done, because they are in the same time zone (effectively). It happens to work out here because we convert to/from the system timezone, so the arithmetic here should cancel itself out, but I'm not sure if that will actually be the case around daylight savings time boundaries.

(2) If there is a calendar given, JDBC is frustratingly vague about the intention of the parameter. It seems most sources interpret it as, "treat the date value as a local date (without reference to any timezone) and give me a timestamp that would display that date (at midnight) in the given timezone". In that case, again, I don't see why the system timezone should come into play.

Reply:
Hi,
initially, you have a timestamp at database which does not have an associated time zone. It can be in any of them.

- when you invoke the method [getTimestamp(int columnIndex)](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#getTimestamp-int-), you are saying that the timestamp you are getting is at the default timezone from the JVM.  

- when you invoke the method [getTimestamp(int columnIndex, Calendar cal)](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#getTimestamp-int-java.util.Calendar-), you are saying that the timestamp you are getting is at the timezone specified at input `cal` parameter. 

Check the example I attached before. You have '2005-06-29 19:19:41' at database.

- your JVM is in `America/Los_Angeles` and you invoke the [getTimestamp(int columnIndex)](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#getTimestamp-int-). At `ArrowFlightJdbcTimeStampVectorAccessor` you receives that exact timestamp in UTC. But, you have to translate it to the same timestamp value but at `America/Los_Angeles` (the default JVM timezone). 

  If you print the obtained timestamp, you will see the same value that was at database. 
  Something like:
`2005-06-29 19:19:41.0`

- your JVM is in `America/Los_Angeles` and you invoke the [getTimestamp(int columnIndex, Calendar cal)](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#getTimestamp-int-java.util.Calendar-) specifying `Asia/Tokyo`. Same as before, you receive the timestamp in UTC at `ArrowFlightJdbcTimeStampVectorAccessor`, but now you have to put that same timestamp in `Asia/Tokyo` time zone, because is the one specified at the `getTimestamp` invokation.
 
  When you print the obtained timestamp, you will see this:
`2005-06-29 03:19:41.0`

  This is because the timestamp is printed following your JVM timezone (America/Los_Angeles). But note that it is the expected value because you retrieved the timestamp using `getTimestamp(index, Asia/Tokyo-calendar)`. So, you specified that the timestamp  '2005-06-29 19:19:41'  is at the time zone `Asia/Tokyo`. 

   If you check the difference, there are 16 hours less, because `America/Los_Angeles` is at GMT-7 and `Asia/Tokyo` is at UTC+9. This is why the timestamp is printed as `2005-06-29 03:19:41.0`, `America/Los_Angeles` value, which is '2005-06-29 19:19:41.0' at `Asia/Tokyo`

Reply:
I'm talking about Date here still.

(1) Arrow's Date is defined in terms of the Unix epoch (i.e. UTC), and (2) java.sql.Date is also defined in terms of GMT. So it doesn't seem like the system timezone should be relevant.

Reply:
I'm also not really concerned about what it prints, but what the actual represented value is

Reply:
Yeesh, system timezone appears to be a convention (e.g. [this StackOverflow answer](https://stackoverflow.com/questions/9202857/timezones-in-sql-date-vs-java-sql-date)) but doesn't appear to be formally specified. I still think that, given Arrow actually specifies the epoch for its values, introducing the system timezone merely to be consistent with databases that appear not to store this information is strictly wrong.

Reply:
I suppose it could depend on how you want to read this sentence...

> To conform with the definition of SQL DATE, the millisecond values wrapped by a java.sql.Date instance must be 'normalized' by setting the hours, minutes, seconds, and milliseconds to zero in the particular time zone with which the instance is associated.

Reply:
ResultSet#getDate never specifies the time zone when a calendar is not provided, so I guess the question is whether we should use the underlying value's time zone or the system time zone

Reply:
Hi, I will be out for a few days. I will try to answer questions when I return. 
In the meantime, I would like to comment:
> Yeesh, system timezone appears to be a convention (e.g. [this StackOverflow answer](https://stackoverflow.com/questions/9202857/timezones-in-sql-date-vs-java-sql-date)) but doesn't appear to be formally specified. I still think that, given Arrow actually specifies the epoch for its values, introducing the system timezone merely to be consistent with databases that appear not to store this information is strictly wrong.

It is true that documentation is not 100% clear, but it is also true that other JDBC drivers from important vendors work like that, using the default JVM timezone. So, a generic JDBC client will expect the data on that way. 

> I'm talking about Date here still.
(1) Arrow's Date is defined in terms of the Unix epoch (i.e. UTC), and (2) java.sql.Date is also defined in terms of GMT. So it doesn't seem like the system timezone should be relevant.

I will check again the code. But the problem is the same. The `java.sql.Date` has also hours, minutes and seconds. So, the Arrow JDBC driver retrieves the value in UTC, ok, but when a new `Date` is created it is done on the default JVM timezone. So, it can generate a different date than the expected. This is why a similar translation than the applied with `Timestamp` is required.